Topics/Tech_Stacks/Peewee_and_Pydantic_models.md

```diff
@@ -0,0 +1,257 @@
+
+```python
+class User(Model):
+    username = CharField(unique=True)
+    email = CharField(unique=True)
+```
+This enforces uniqueness for both `username` and `email`.
+
+As previously stated, keys and constraints help enforce relationships between relational database tables. There are a lot of different upsides to designing good databases as it could ensure data integrity, and improve query performance in schemas. Thus, it is worthwhile to spend time designing the relational database tables within your application! What I have covered here is just the tip of the iceberg! You can learn more about keys [here](https://docs.peewee-orm.com/en/latest/peewee/models.html#primary-keys-composite-keys-and-other-tricks).
+
+## Defining Default Values for a Field
+Sometimes it makes more sense to have a default value. For example, the Event model above defines an Event happening.
+
+```python
+class Event(Model):
+    event_date = DateField()
+    event_time = TimeField()
+```
+
+But what if the event has no event_time as it does not start at a predetermined time but is over the course of a day or a few days. Take for example defining Halloween as an event. The TimeField should default as None, as it happens throughout the day:
+
+```python
+class Event(Model):
+    event_date = DateField()
+    event_time = TimeField(default=None)
+```
+
+As shown above, we use the `default` keyword and set it as None.
+
+## Using Pydantic for Data Validation
+Next, create a Pydantic model for data validation. In a file like schemas.py:
+
+```python
+from pydantic import BaseModel
+
+class PersonSchema(BaseModel):
+    name: str
+    age: int
+```
+
+Here, `PersonSchema` mirrors is exactly the same as the fields of the Peewee model. After, this Pydantic will ensure that the data you receive or send adheres to the specified structure!
+
+## CRUD Operations with Peewee and Pydantic Validation
+Now, let's perform CRUD (Create, Read, Update, Delete) operations using Peewee. In a new file, e.g., crud_operations.py:
+
+```python
+from models import Person, db
+from schemas import PersonSchema
+
+# Create
+def create_person(person_data) -> PersonSchema:
+    # creates instance of the PersonSchema, using double astricks syntax to unpack contents
+    person_schema = PersonSchema(**person_data)
+    # convert pydantic model into a dictionary, ensuring data conforms to expected structure of Pydantic model
+    validated_data = person_schema.dict()
+    person = Person.create(**validated_data)
+    return PersonSchema(**person.__dict__)  # Return as PersonSchema instance
+
+# Read
+def get_person(person_id) -> PersonSchema:
+    person = Person.get_by_id(person_id)
+    return PersonSchema(**person.__dict__)  # Return as PersonSchema instance
+
+# Update
+def update_person(person_id, new_data):
+    person = Person.get_by_id(person_id)
+    
+    # Validate and update data
+    person_data = PersonSchema.from_orm(person).dict()
+    updated_data = {**person_data, **new_data}
```

Comment:
> updated_data = {**person_data, **new_data}

I'd be interested to see how this notation for constructing a dictionary ends up looking once initialized. Maybe it would be a good idea to add a comment here with an example of one of these dictionaries?